### PR TITLE
Migrate over protocols.

### PR DESCRIPTION
**Pull request by ckipp01**
This has been something I've been meaning to work on for quite some time. We still have quite a bit of duplicated code around our extensions. This unifies that. I'll have prs following in coc-metals and the vs code extension as well.

### CODE REVIEW

File: src/interfaces/MetalsInputBox.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  >("metals/inputBox");
+}
+
+export interface MetalsInputBoxResult {
+  value?: string;
+  cancelled?: boolean;
+}
+
+/**
+ * The below is ported directly from VS Code:
```
Comment (ckipp01) [inline src/interfaces/MetalsInputBox.ts:30]: This is a bit unfortunate, but since this is tied to the vs code api, I don't see a way around it. This won't really help the vs code extension much, but I have all of this duplicated in coc-metals. I was opposed to moving this over at first, but then realized if this api changes, then the Metals implementation needs to change anyways. So, I figured this was ok to port over. I was also going to do the same thing with decorations, but that goes _way_ deeper than this, and I've have to port over like 3 times this amount of stuff, which I don't think is a good idea. That's way the decoration protocol isn't found in here.

File: src/interfaces/ClientCommands.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/**
+ * These are all of the client commands that Metals supports. Note that not support
+ * may vary based on the `InitializationSettings` the client sets.
+ *
+ *  - https://scalameta.org/metals/docs/editors/new-editor.html#metals-client-commands
+ */
+export enum ClientCommands {
```
Comment (gabro) [inline src/interfaces/ClientCommands.ts:7]: nitpick: `enum` are a discouraged feature of TypeScript: they are one of the few features that has a runtime behavior and the community is essentially moving away from them.

I would recommend using a literal strings union or an object instead.

Literal strings union:

```
type ClientCommands =
| "metals-echo-command"
| "metals-diagnostics-focus"
// ...
```

This has the minor downsides of not having pretty names (but it's the simplest option in my opinion)

If you go with an object, you can get pretty names:


```
const clientCommands = {
  EchoCommand: "metals-echo-command",
  FocusDiagnostics: "metals-diagnostics-focus",
  // ...
} as const
type ClientCommands = typeof clientCommands
```

`as const` is there to prevent TS from widening the value types to `string` and preserving the literal types instead.

This requires a bit more incantation for exhaustive matching, but it's not too bad if we have `fp-ts` as a dependency. Not sure it's worth the effort though.

File: src/interfaces/MetalsInputBox.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+   */
+  validateInput?(
+    value: string
+  ): string | undefined | null | Thenable<string | undefined | null>;
+}
+
+/**
+ * Represents options for a specific decoration in a [decoration set](#TextEditorDecorationType).
+ */
+export interface DecorationOptions {
```
Comment (gabro) [inline src/interfaces/MetalsInputBox.ts:82]: why is this here in the `MetalsInputBox` file?

File: src/interfaces/MetalsInputBox.ts (hunk 1 above)
Comment (gabro) [reply src/interfaces/MetalsInputBox.ts:30]: I see the problem. One way to work around the issue is to define the protocol interface here in an "agnostic way" and then to have specific implementations of the protocol in the clients.

Methods in metals-languageclient that need to use an input box can require the implementation from the client in a "manual typeclass" fashion.

Does it make any sense?

File: src/interfaces/ServerCommands.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/**
+ * These are all of the server commands that Metals supports. Note that not support
+ * may vary based on the `InitializationSettings` the client sets.
+ *
+ *  - https://scalameta.org/metals/docs/editors/new-editor.html#metals-server-commands
+ */
+export enum ServerCommands {
```
Comment (gabro) [inline src/interfaces/ServerCommands.ts:7]: Same as above

File: src/interfaces/ClientCommands.ts (hunk 2 above)
Comment (gabro) [reply src/interfaces/ClientCommands.ts:7]: ah sorry, the first case only defines the types and not the values, but it's still valid

Comment (gabro) [reply src/interfaces/ClientCommands.ts:7]: No, actually, changed my mind: the second case is best if we want to have exhaustive command handling

Comment (ckipp01) [reply src/interfaces/ClientCommands.ts:7]: Gotcha. That makes sense. I originally had an object, then switched, and was overall just unsure. So I'll switch this back to an object 👍

Comment (gabro) [reply src/interfaces/ClientCommands.ts:7]: Cool, see also my comment in metals-vscode on how to handle the object exhaustively

File: src/interfaces/MetalsInputBox.ts (hunk 3 above)
Comment (ckipp01) [reply src/interfaces/MetalsInputBox.ts:82]: Ah good catch. It shouldn't be. I think I had them in the same file in coc-metals and when I copied them over here I copied to much. Removed.

File: src/interfaces/ServerCommands.ts (hunk 4 above)
Comment (ckipp01) [reply src/interfaces/ServerCommands.ts:7]: 👍 Changed to an object.

File: src/interfaces/MetalsInputBox.ts (hunk 1 above)
Comment (ckipp01) [reply src/interfaces/MetalsInputBox.ts:30]: The idea makes sense yes. How to implement this in TS alludes me a bit. Could you give me an example?

Comment (gabro) [reply src/interfaces/MetalsInputBox.ts:26]: Nothing fancy at all:

```ts
// in metals-languageclient

// the "typeclass" definition
interface MetalsInputBox {
  // or something less lazy than this :-P
  function askStuffToUser(message: string): Promise<string>;
}

// some method require the typeclass instance
function someMethodThatAsksThings(inputBox: MetalsInputBox): (message: string) => Promise<unknown> {
  return message => inputBox.askStuffToUser(message).then(...)
}
```

```ts
// in metals-vscode

// instance definition
const inputBox: MetalsInputBox = {
  askStuffToUser: message => {
    // implementation using VSCode API
  }
}

// "bind" the method to the typeclass (optional, but may be useful if invoked a lot)
const askThings = someMethodThatAsksThings(inputBox)

// ...

// use the bound method
askThings("hey, what's your name?").then(...)
```

Comment (ckipp01) [reply src/interfaces/MetalsInputBox.ts:26]: 🤔 I guess I'm struggling with seeing the benifit of this, and the more I think of it, the more I'm actually thinking that porting this into here does make sense. Mainly, I don't think this extra layer of abstraction will gain us anything. For example, this is only used in the following two places right:

In VS Code
```ts
    client.onRequest(MetalsInputBox.type, (options, requestToken) => {
      return window.showInputBox(options, requestToken).then((result) => {
        if (result === undefined) {
          return { cancelled: true };
        } else {
          return { value: result };
        }
      });
    });
```

In coc-metals
```
    client.onRequest(MetalsInputBox.type, (options: InputBoxOptions) => {
      return workspace.callAsync<string>("input", [
        `${options.prompt}: `,
        options.value ? options.value : "",
      ]).then((response) => {
        if (response.trim() === "") {
          return { cancelled: true };
        } else {
          return { value: response };
        }
      })
    });
```

So we would basically create, bind, and use the instance once. I could see this being useful if we used it all over the place, but we don't. Also, if we would do the typeclass approach, I'd still need to define an interface in coc-metals for `InputBoxOptions` which is used.  I sort of think this is just a result of us exactly using `InputBoxOptions` from the VS Code api, and it actually should be part of the shared library if it's part of the protocol.

Thoughts?

Comment (ckipp01) [reply src/interfaces/MetalsInputBox.ts:26]: Basically just adding the following to the namespace:

```ts
  export const handleInput = function (result: string | undefined): MetalsInputBoxResult {
    if (result === undefined || result.trim() === "") {
      return {cancelled: true};
    } else {
      return {value: result};
    }
  }
```

Which would then allow for:

VS Code
```ts
    client.onRequest(MetalsInputBox.type, (options, requestToken) => {
      return window
        .showInputBox(options, requestToken)
        .then(MetalsInputBox.handleInput);
    });
```

coc-metals

```ts
InputBox.type, (options: InputBoxOptions) => {
      return workspace
        .callAsync<string>("input", [
          `${options.prompt}: `,
          options.value ? options.value : "",
        ])
        .then(MetalsInputBox.handleInput);
    });
```

**Comment by ckipp01** [comment]
I sort of feel ok merging this as is, unless you feel strongly against it @gabro.

**Comment by gabro** [comment]
Go ahead, absolutely. I haven't had time to re-review this, but it was good already

**Comment by ckipp01** [comment]
> Go ahead, absolutely. I haven't had time to re-review this, but it was good already

Cool cool. I left a bunch more comments. So whenever you have time. Feel free to go back through them. If I need to revisit and make some changes, I'm more than happy to. 👍 As always, appreciate the input.
